[12.0][FIX] fix product.template related fields

huguesdk opened the pull request:
make `product.template.product_variant_id` a stored computed field to avoid the following error when setting the `print_category_id` or `to_print` related fields:

> non-stored field product.template.product_variant_id cannot be searched.

huguesdk committed:
[FIX] fix product.template related fields

make product.template.product_variant_id a stored computed field to
avoid the following error when setting the print_category_id or to_print
related fields: "non-stored field product.template.product_variant_id
cannot be searched."

legalsylvain commented:
Hello @huguesdk ! Je vais laisser la PR ouverte. la cible c'est désormais la V16, ou j'ai mis le module dans l'OCA.
ça joue ? 
https://github.com/OCA/product-attribute/tree/16.0/product_print_category#product---print-categories

huguesdk commented:
ok, fine for me. i’ll test and create a similar pr for version 16 on the oca repo then.

carmenbianca committed:
[FIX] product_print_category: Improve computation of product_variant_id for inactive templates

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
@huguesdk added a patch to your code, internal task T12969

carmenbianca commented:
i also made a pr here : https://github.com/odoo/odoo/pull/181811

carmenbianca committed:
[IMP] product_print_category: Recompute instead of overriding the compute method

Overriding methods is bad practice. By simply recomputing at the correct
stage in the write method, we achieve effectively the same.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
Upstream PR likely won't be merged. I added a new commit ~~d5d7e05~~ 61ae3f8 with a new insight to achieve the same result without overriding the compute method.

huguesdk commented on product_print_category/models/product_template.py:
```diff
@@ -18,33 +18,14 @@ class ProductTemplate(models.Model):
+        # Recompute product_variant_id to circumvent a bug where
+        # product_variant_ids cannot be accessed normally (empty recordset)
+        # because they are archived at time of computation.
+        if "active" in vals:
+            self._compute_product_variant_id()
```
i don’t understand how this fixes the problem where an archived product template is duplicated and has no variants. `_compute_product_variant_id()` accesses `product_variant_ids`, which won’t contain archived variants. or am i missing something?

also, do you know why the test is failing?

carmenbianca replied:
Fixed the test.

Between trying to upstream this and patching it here, I've also had a wrong assumption about how things work in Odoo 12.

In Odoo 12:

- When copying a product.template which has product variants, the copied result does NOT have any product variants.

In Odoo 15 (and maybe earlier):

- When copying a product.template which has product variants, the `product.template.attribute.value`s (PTAVs) from which those `product.product`s were generated are copied. Then, during the `create` method of the product.template, new `product.product`s are created from those PTAVs.

---

To answer your question:

Indeed, when a product.template is inactive, `_compute_product_variant_id()` sets False, because `product_variant_ids` only contains archived products and will thus be an empty recordset. This appears to be by design, so we match upstream behaviour here.

When a template is (re)activated, instead of reactivating its product variants, **new** `product.product`s are created from the PTAVs. Then by running our compute method after that has happened, `product_variant_id` is correctly populated.

But by running that compute method again when the template (and thus its also its variants) is unarchived, we get a sensible result. This matches upstream behaviour precisely.

huguesdk replied:
nicely done, thanks for the explanation!

carmenbianca committed:
[IMP] product_print_category: Recompute by declaring additional dependant field

This should have been obvious from the start...

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
@huguesdk i fixed this a final time with a solution that is way more obvious. the tests succeed on my machine.
